explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,260 @@
+WARNING: Logging before InitGoogleLogging() is written to STDERR
+E20240928 05:30:42.037626 1419008 sample_reader.cc:101] Cannot find binary with buildid 48b4b9e5f4dc4b07680bfc67d333b5893fbe0162
+I20240928 05:30:42.117508 1419008 symbol_map.cc:477] Adding loadable exec segment: offset=457000 vaddr=457000
+```
+
+Nevertheless, you should inspect its output and also make sure to check the `$?` shell variable to make sure that it has finished successfully.  However, even if `create_gcov` finishes with exit `0` the generated file `openssl-nonpgo.gcov` might not be valid.  It is also a good idea to use the `dump_gcov` tool on it and make sure that it actually contains valid information.  For example:
+
+```bash
+$ dump_gcov openssl-nonpgo.gcov
+cpu_get_tb_cpu_state total:320843603 head:8557544
+  2: 8442523
+  7: 8442518
+  7.1: 8442518
+  7.4: 9357851
+  8: 9357851
+  10: 9357854
+  19: 0
+  21: 0
+[...] many more lines [...]
+```
+
+If the command does not print anything, there is a problem with your `gcov` file.
+
+If everything went well, we can now use the `gcov` file and feed it back to GCC when recompiling OpenSSL.
+
+### Rebuilding OpenSSL with PGO
+
+We have everything we need to rebuild OpenSSL and make use of our profile data.  The most important thing now is to set the `CFLAGS` environment variable properly so that GCC can find and use the profile we generated in the previous step.
+
+How you perform this `CFLAGS` adjustment depends on how you built the software in the first place, so we won't cover this part here.  The resulting `CFLAGS` variable should have the following GCC option, though:
+
+```
+-fauto-profile=/path/to/openssl-nonpgo.gcov
+```
+
+Make sure to adjust the path to the `openssl-nonpgo.gcov` file.
+
+Also note that enabling PGO will make GCC automatically enable several optimization passes that are usually disabled, even when using `-O2`.  This can lead to new warnings/errors in the code, which might require either the code to be adjusted or the use of extra compilation flags to suppress these warnings.
+
+After the software is built, you can install this new version and run the benchmark tests again to compare results.
+
+### Our results
+
+At first sight it might appear that our results don't have much to show for themselves, but they are actually very interesting and allow us to discuss some more things about PGO.
+
+First, it is not a silver bullet.  It will not work for every scenario, and there might be cases when it is not justified to use it.  For example, if you are dealing with a very generic program that doesn't have a single representative workload.  PGO is great when applied to specialized cases, but may negatively impact the performance of programs that can be used in a variety of ways.
```

Comment:
```suggestion
First, it is not a "cure all".  It will not work for every scenario, and there might be cases when it is not justified to use it.  For example, if you are dealing with a very generic program that doesn't have a single representative workload.  PGO is great when applied to specialized cases, but may negatively impact the performance of programs that can be used in a variety of ways.
```

Reply:
"cure all" is more likely to carry your meaning than "silver bullet" across language divides -- tbh we could remove that part completely and just say "first, it will not work for..." to avoid the need for idioms, but it's not a blocker.

Reply:
Thanks.  I decided to simplify and remove the "silver bullet" part as suggested.